node/bindnode/node.go

```diff
@@ -82,7 +84,38 @@ func (w *_node) Representation() datamodel.Node {
 }
 
 func (w *_node) Kind() datamodel.Kind {
-	return w.schemaType.TypeKind().ActsLike()
+	return actualKind(w.schemaType)
+}
+
+func compatibleKind(schemaType schema.Type, kind datamodel.Kind) error {
+	switch sch := schemaType.(type) {
+	case *schema.TypeAny:
+		return nil
+	default:
+		actual := actualKind(sch)
+		if actual == kind {
+			return nil
+		}
+		methodName := ""
+		if pc, _, _, ok := runtime.Caller(1); ok {
+			if fn := runtime.FuncForPC(pc); fn != nil {
+				methodName = fn.Name()
+				// Go from "pkg/path.Type.Method" to just "Method".
+				methodName = methodName[strings.LastIndexByte(methodName, '.')+1:]
+			}
+		}
```

Comment:
This is probably really nice and we probably should've done it ages ago.

But just to talk it through outloud: is this at all significantly expensive in execution time?  I don't have a good intuition for how much those `runtime.*` calls cost.

Reply:
On second thought: Nevermind.  I retract my interest to only academic.  Because this is only used in error paths, and specifically, an error path that should always be avoidable by code that bothers to check the `.Kind()` (which is about the cheapest operation one can make): it is not necessary to care about the possibility of a slow error reporting path.

Reply:
Correct, this does not slow down the common path at all. The cost of inspecting the call stack is noticeable; I seem to recall it's in the order of microseconds. Not something you'd want to do on every method call, but entirely reasonable when constructing an error.